pbkstore.go

```diff
@@ -16,62 +16,107 @@
+)
+
+// PiggyBackStore store a piggyback data. When swim ping, ack or indirect-ack,
+// swim need to get PiggyBackData from the PiggyBackStore and send it with ping, ack or indirect-ack.
+
+// The priority is incremented each time when the data is queried.
+// The lower the priority, the faster the queried.
 // The data has a member status, such as alive, suspected or dead.
-type piggyBackData struct {
-	data       []byte
-	localCount int
-}
 
-// PiggyBackStore store a piggyback data. When we ping, ack or indirect-ack,
-// get PiggyBackData from the PiggyBackStore and send it with ping, ack or indirect-ack.
-type PiggyBackStore interface {
+type PBkStore interface {
 	Len() int
 	Push(b []byte) error
 	Get() ([]byte, error)
 	IsEmpty() bool
 }
 
-// PiggyBackPriorityStore stores piggyback data and returns data with small local count.
-type PiggyBackPriorityStore struct {
-	list    []piggyBackData
-	maxSize int
-	lock    sync.RWMutex
+// PiggyBackStore stores piggyback data in the priority queue and returns data with smallest local count.
+type PriorityPBStore struct {
+	q             PriorityQueue
+	maxLocalCount int
+	lock          sync.RWMutex
 }
 
-func NewPiggyBackPriorityStore(maxSize int) *PiggyBackPriorityStore {
-	return &PiggyBackPriorityStore{
-		list:    make([]piggyBackData, 0),
-		maxSize: maxSize,
-		lock:    sync.RWMutex{},
+// macLocalCount is the max priority value
+func NewPriorityPBStore(maxLocalCount int) *PriorityPBStore {
+	return &PriorityPBStore{
+		q:             make(PriorityQueue, 0),
+		maxLocalCount: maxLocalCount,
+		lock:          sync.RWMutex{},
 	}
 }
 
 // Return current size of data
-func (p *PiggyBackPriorityStore) Len() int {
+func (p *PriorityPBStore) Len() int {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	return len(p.list)
+	return p.q.Len()
 }
 
 // Enqueue []byte into list.
 // Initially, set the local count to zero.
 // If the queue size is max, delete the data with the highest localcount and insert it.
-func (p *PiggyBackPriorityStore) Push(b []byte) error {
+func (p *PriorityPBStore) Push(pbkData []byte) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	item := &Item{
+		value:    pbkData,
+		priority: InitialPriority,
+	}
+
+	heap.Push(&p.q, item)
 	return nil
 }
 
 // Return the piggyback data with the smallest local count in the list,
 // increment the local count and sort it again, not delete the data.
-func (p *PiggyBackPriorityStore) Get() ([]byte, error) {
-	return nil, nil
+func (p *PriorityPBStore) Get() ([]byte, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// Check empty
+	if len(p.q) == 0 {
+		return nil, ErrStoreEmpty
+	}
+
+	// Pop from queue
+	item := heap.Pop(&p.q).(*Item)
+	b, ok := item.value.([]byte)
+	if !ok {
+		return nil, ErrPopInvalidType
+	}
+
+	// If an item has been retrieved by maxPriority, remove it.
+	// If not, push it again after increment priority
+	item.priority = item.priority + 1
+	if item.priority < p.maxLocalCount {
+		p.q.Push(item)
+	}
+
```

Comment:
how about remove error return type from `Get`, `Push`?

`Push` has no possibility to return error
and although `Get` can return nil value if Store is empty, I think these nil cases can be handled from outside

Reply:
Returning error from push can be removed, but if we remove returning error from `Get`, we do know why we fail to get data from `Get` function. So i think `Get` should return error

Reply:
ok sounds reasonable